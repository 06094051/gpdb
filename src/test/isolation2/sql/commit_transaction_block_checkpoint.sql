-- TEST 1: block checkpoint on segments

-- pause the 2PC after setting inCommit flag
! gpfaultinjector -f twophase_transaction_commit_prepared -m async -y suspend -o 0 -H ALL -r primary;

-- trigger a 2PC, and it will block at commit;
2: checkpoint;
2: begin;
2: create table t_commit_transaction_block_checkpoint (c int) distributed by (c);
2&: commit;

-- do checkpoint on dbid 3 (segment 1) in utility mode, and it should block
3U&: checkpoint;

-- resume the 2PC after setting inCommit flag
! gpfaultinjector -f twophase_transaction_commit_prepared -m async -y reset -o 0 -H ALL -r primary;
2<:
3U<:

-- TEST 2: block checkpoint on master

-- pause the CommitTransaction right before persistent table cleanup after
-- notifyCommittedDtxTransaction()
! gpfaultinjector -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y suspend -o 0 -s 1;

-- trigger a 2PC, and it will block at commit;
2: checkpoint;
2: begin;
2: drop table t_commit_transaction_block_checkpoint;
2&: commit;

-- do checkpoint on dbid 1 (master) in utility mode, and it should block
1U&: checkpoint;

-- resume the 2PC
! gpfaultinjector -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y reset -o 0 -s 1;
2<:
1U<:

