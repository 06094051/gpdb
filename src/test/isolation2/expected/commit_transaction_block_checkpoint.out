-- TEST 1: block checkpoint on segments

-- pause the 2PC after setting inCommit flag
! gpfaultinjector -f twophase_transaction_commit_prepared -m async -y suspend -o 0 -H ALL -r primary;
20170303:00:02:12:067930 gpfaultinjector:-[INFO]:-Starting gpfaultinjector with args: -f twophase_transaction_commit_prepared -m async -y suspend -o 0 -H ALL -r primary
20170303:00:02:12:067930 gpfaultinjector:-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170303:00:02:12:067930 gpfaultinjector:-[INFO]:-Injecting fault on content=1:dbid=3:mode=s:status=u
20170303:00:02:12:067930 gpfaultinjector:-[INFO]:-Injecting fault on content=2:dbid=4:mode=s:status=u
20170303:00:02:12:067930 gpfaultinjector:-[INFO]:-DONE


-- trigger a 2PC, and it will block at commit;
2: checkpoint;
CHECKPOINT
2: begin;
BEGIN
2: create table t_commit_transaction_block_checkpoint (c int) distributed by (c);
CREATE
2&: commit;  <waiting ...>

-- do checkpoint on dbid 3 (segment 1) in utility mode, and it should block
3U&: checkpoint;  <waiting ...>

-- resume the 2PC after setting inCommit flag
! gpfaultinjector -f twophase_transaction_commit_prepared -m async -y reset -o 0 -H ALL -r primary;
20170303:00:02:14:067954 gpfaultinjector:-[INFO]:-Starting gpfaultinjector with args: -f twophase_transaction_commit_prepared -m async -y reset -o 0 -H ALL -r primary
20170303:00:02:14:067954 gpfaultinjector:-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170303:00:02:14:067954 gpfaultinjector:-[INFO]:-Injecting fault on content=1:dbid=3:mode=s:status=u
20170303:00:02:14:067954 gpfaultinjector:-[INFO]:-Injecting fault on content=2:dbid=4:mode=s:status=u
20170303:00:02:14:067954 gpfaultinjector:-[INFO]:-DONE

2<:  <... completed>
COMMIT
3U<:  <... completed>
CHECKPOINT

-- TEST 2: block checkpoint on master

-- pause the CommitTransaction right before persistent table cleanup after
-- notifyCommittedDtxTransaction()
! gpfaultinjector -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y suspend -o 0 -s 1;
20170303:00:02:15:067973 gpfaultinjector:-[INFO]:-Starting gpfaultinjector with args: -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y suspend -o 0 -s 1
20170303:00:02:15:067973 gpfaultinjector:-[INFO]:-Injecting fault on content=-1:dbid=1:mode=s:status=u
20170303:00:02:15:067973 gpfaultinjector:-[INFO]:-DONE


-- trigger a 2PC, and it will block at commit;
2: checkpoint;
CHECKPOINT
2: begin;
BEGIN
2: drop table t_commit_transaction_block_checkpoint;
DROP
2&: commit;  <waiting ...>

-- do checkpoint on dbid 1 (master) in utility mode, and it should block
1U&: checkpoint;  <waiting ...>

-- resume the 2PC
! gpfaultinjector -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y reset -o 0 -s 1;
20170303:00:02:16:067988 gpfaultinjector:-[INFO]:-Starting gpfaultinjector with args: -f transaction_commit_pass1_from_drop_in_memory_to_drop_pending -m async -y reset -o 0 -s 1
20170303:00:02:16:067988 gpfaultinjector:-[INFO]:-Injecting fault on content=-1:dbid=1:mode=s:status=u
20170303:00:02:16:067988 gpfaultinjector:-[INFO]:-DONE

2<:  <... completed>
COMMIT
1U<:  <... completed>
CHECKPOINT

